src/game/client/gui/controlbar/controlbar.cpp

```diff
@@ -83,27 +265,229 @@ bool CommandButton::Is_Valid_Relationship_Target(Relationship r) const
 
     if (r) {
         if (r == ALLIES) {
-            options = 4;
+            options = COMMAND_OPTION_NEED_TARGET_ALLY_OBJECT;
         } else if (r == NEUTRAL) {
-            options = 2;
+            options = COMMAND_OPTION_NEED_TARGET_NEUTRAL_OBJECT;
         }
     } else {
-        options = 1;
+        options = COMMAND_OPTION_NEED_TARGET_ENEMY_OBJECT;
     }
 
     return (options & m_options) != 0;
 }
 
-const CommandButton *CommandSet::Get_Command_Button(int button) const
+void CommandButton::Parse_Command(INI *ini, void *formal, void *store, const void *user_data)
 {
-    const CommandButton *c;
-    if (g_theGameLogic != nullptr && g_theGameLogic->Find_Control_Bar_Override(m_name, button, c)) {
-        return c;
+    const char *str = ini->Get_Next_Token();
+
+    for (int i = 0; s_theGuiCommandNames[i] != nullptr; i++) {
+        if (strcasecmp(s_theGuiCommandNames[i], str) == 0) {
+            *reinterpret_cast<GUICommand *>(store) = static_cast<GUICommand>(i);
+            return;
+        }
     }
 
-    return m_command[button];
+    throw CODE_06;
+}
+
+bool CommandButton::Is_Valid_To_Use_On(Object *obj, const Object *obj2, const Coord3D *pos, CommandSourceType type) const
+{
+    if (m_upgradeTemplate != nullptr) {
+        ProductionUpdateInterface *interface = obj->Get_Production_Update_Interface();
+
+        if (interface != nullptr) {
+            for (const ProductionEntry *entry = interface->First_Production(); entry;
+                 entry = interface->Next_Production(entry)) {
+                if (entry->Get_Production_Object() != nullptr) {
+                    return false;
+                }
+            }
+
+            return obj->Affected_By_Upgrade(m_upgradeTemplate) && !obj->Has_Upgrade(m_upgradeTemplate);
+        } else {
+            return false;
+        }
+    } else {
+        if ((m_options
+                & (COMMAND_OPTION_NEED_TARGET_ALLY_OBJECT | COMMAND_OPTION_NEED_TARGET_NEUTRAL_OBJECT
```

Comment:
Same as the other file. I suggest to create an alias for `COMMAND_OPTION_NEED_TARGET_ALLY_OBJECT | COMMAND_OPTION_NEED_TARGET_NEUTRAL_OBJECT | COMMAND_OPTION_NEED_TARGET_ENEMY_OBJECT` to break these conditions down.

Alias can be an enum value or constexpr value in this function.

```c++
COMMAND_OPTION_NEED_TARGET_OBJECT = COMMAND_OPTION_NEED_TARGET_ALLY_OBJECT | COMMAND_OPTION_NEED_TARGET_NEUTRAL_OBJECT | COMMAND_OPTION_NEED_TARGET_ENEMY_OBJECT;
```
or
```c++
constexpr auto any_target_object = COMMAND_OPTION_NEED_TARGET_ALLY_OBJECT | COMMAND_OPTION_NEED_TARGET_NEUTRAL_OBJECT | COMMAND_OPTION_NEED_TARGET_ENEMY_OBJECT;
```

Reply:
fixed.